internal/scripts/release/crates_io_publish_script.sh

```diff
@@ -0,0 +1,183 @@
+)
+
+if [[ "$#" -eq 0 ]]; then
+    echo -e "${C_RED}ERROR:${C_OFF} No arguments provided. Try 'help' for options."
+    exit 1
+fi
+
+while (( "$#" )); do
+    case "$1" in
+        "dry-run")
+            DO_DRY_RUN=true
+            shift 1
+            ;;
+        "publish")
+            DO_PUBLISH=true
+            shift 1
+            ;;
+        "sanity-checks")
+            DO_SANITY_CHECKS=true
+            shift 1
+            ;;
+        "help")
+            echo -e "Script to publish to crates.io"
+            echo -e ""
+            echo -e "Usage: ${C_GREEN}$(basename $0)${C_OFF} publish"
+            echo -e "Options:"
+            echo -e "    dry-run            Simulate publishing to crates.io"
+            echo -e "                       Only works with Rust >= 1.90"
+            echo -e "    publish            Publish to crates.io"
+            echo -e "    sanity-checks      Sanity checks for cyclic dependencies and new crates"
+            echo -e ""
+            exit 0
+            ;;
+        *)
+            echo -e "${C_RED}ERROR:${C_OFF} Invalid argument '$1'. Try 'help' for options."
+            exit 1
+            ;;
+    esac
+done
+
+cd $(git rev-parse --show-toplevel)
+
+sanity_check_completeness() {
+    ICEORYX2_CRATES=$(grep -E '^iceoryx2' Cargo.toml 2>/dev/null | awk '{print $1}' | sort -u)
+
+    for CRATE in ${ICEORYX2_CRATES}; do
+        if [[ " ${CRATES_TO_IGNORE[@]} " =~ " ${CRATE} " ]]; then
+            continue
+        fi
+
+        if [[ " ${CRATES_TO_PUBLISH[@]} " =~ " ${CRATE} " ]]; then
+            continue
+        fi
+
+        echo -c "${C_RED}ERROR:${C_OFF} Crate '$CRATE' is neither in the publish nor ignore list."
+        exit 1
+    done
+}
+
+sanity_check_cyclic_dependencies() {
+    declare -a ALLOWED_CRATE_DEPENDENCIES
+
+    HAS_ERROR=false
+    for CRATE in "${CRATES_TO_PUBLISH[@]}"; do
+        ALLOWED_CRATE_DEPENDENCIES+=(${CRATE})
+
+        CRATE_DEPENDENCIES=$(cargo tree --package "${CRATE}" --depth 1 --prefix none | grep -v '(\*)' | grep '^iceoryx2' | awk '{print $1}' | sort | uniq)
+        for DEP in ${CRATE_DEPENDENCIES}; do
+            if [[ " ${ALLOWED_CRATE_DEPENDENCIES[@]} " =~ " ${DEP} " ]]; then
+                continue
+            else
+                echo -e "${C_RED}ERROR:${C_OFF} ${C_YELLOW}${CRATE}${C_OFF} dependency ${C_BLUE}${DEP}${C_OFF} is out of order for publishing!"
+                HAS_ERROR=true
+            fi
+        done
+    done
+
+    if [[ ${HAS_ERROR} == true ]]; then
+        exit 1
+    fi
+}
+
+sanity_checks() {
+    sanity_check_completeness
+    sanity_check_cyclic_dependencies
+}
+
+dry_run() {
+    EXCLUDE_ARGS=""
+    for CRATE in "${CRATES_TO_IGNORE[@]}"; do
+        EXCLUDE_ARGS+="--exclude $CRATE "
+    done
+    cargo publish --dry-run --workspace ${EXCLUDE_ARGS}
+}
+
+publish() {
+    for CRATE in ${CRATES_TO_PUBLISH[@]}; do
+        echo -e "${C_BLUE}${CRATE}${C_OFF}"
+        cargo publish -p ${CRATE}
+    done
+}
```

Comment:
@elfenpiff @dkroenke shall we also make Rust 1.90 mandatory for the actual publishing?

With Rust 1.90 the publishing of workspaces became stable. It is a bit more "dangerous", since all crates are published which are not explicitly excluded. While there is a check for new crates which start with `iceoryx2`, there is no check for other namings, like `benchmark`.

I would therefore need to make the check for new crates more sophisticated, in order for us to add them to the `CRATES_TO_PUBLISH` or `CRATES_TO_IGNORE` list, or add a comment why we publish the crates separately instead of the whole workspace.